### PR TITLE
Stats: Move Jetpack Stats Widget in WP Dashboard to Own Class

### PR DESCRIPTION
**Pull request by sdixon194**
#### Changes proposed in this Pull Request:

* This PR is part of a larger refactor of our stats module, starting by moving the Jetpack stats widget in the WordPress dashboard into its own class, beginning with removing the widget code from class.jetpack.php.

#### Does this pull request change what data or activity we track or use?

No

#### Testing instructions:

* Switch to this branch, preferably on a site that has some stat data.
* Make sure the stats widget on the WordPress Dashboard (not the Jetpack one), continues to work.
* Check for any regressions

### CODE REVIEW

File: projects/plugins/jetpack/modules/stats/class-jetpack-stats-dashboard-widget.php
Comment (sdixon194) [inline projects/plugins/jetpack/modules/stats/class-jetpack-stats-dashboard-widget.php:72]: PHPCS gives me:

> Overriding WordPress globals is prohibited. Found assignment to $wp_meta_boxes

Don't know if we can ignore this or need to find another solution.

File: projects/plugins/jetpack/modules/stats/class-jetpack-stats-dashboard-widget.php
Comment (sdixon194) [inline projects/plugins/jetpack/modules/stats/class-jetpack-stats-dashboard-widget.php:127]: Gotta figure out how best to escape this.

Comment (kraftbj) [reply projects/plugins/jetpack/modules/stats/class-jetpack-stats-dashboard-widget.php:127]: `esc_html` be fine? We're expecting just a number right?

Comment (sdixon194) [reply projects/plugins/jetpack/modules/stats/class-jetpack-stats-dashboard-widget.php]: D'oh, you're right, I was trying `esc_html_e` but this obviously doesn't need to be translated.

**Comment by sdixon194** [comment]
Noting that this removes any wp dashboard from showing when in offline mode. Currently, in offline mode the widget shows Akismet/Spam blocked #'s (the footer of the dashboard widgets) - however moving this to modules/stats.php means it never loads at all in offline mode.

**Comment by kraftbj** [comment]
> Noting that this removes any wp dashboard from showing when in offline mode. Currently, in offline mode the widget shows Akismet/Spam blocked #'s (the footer of the dashboard widgets) - however moving this to modules/stats.php means it never loads at all in offline mode.

I think that's fine.

File: projects/plugins/jetpack/modules/stats/class-jetpack-stats-dashboard-widget.php
Comment (kraftbj) [reply projects/plugins/jetpack/modules/stats/class-jetpack-stats-dashboard-widget.php:72]: Looking, what about using `add_meta_box`.

`dashboard` would be `$screen->id`, `normal` would be `$context`, `core` would be `priority`, etc.

Looking in the commit history, https://github.com/Automattic/jetpack/pull/1716 ... doesn't really tell us much about what's going on.

**Comment by kraftbj** [comment]
Question: Could we make this a class of itself at the jetpack level, instead of within stats? Looks like this was added to the Jetapack class originally so other things could add to it (protect, akismet), so what if instead of putting it within stats, it's top level.

I haven't tested, but I'm thinking of the use case where stats are disabled but protect is enabled. I _presume_ it would display the protect stats, but without the stats-module stats.

**Comment by sdixon194** [comment]
> Question: Could we make this a class of itself at the jetpack level, instead of within stats? Looks like this was added to the Jetapack class originally so other things could add to it (protect, akismet), so what if instead of putting it within stats, it's top level.

Not a bad idea. So we make a general "Jetpack Dashboard Widget" class that we can use to load whatever we want, ie akismet spam blocked, stats, etc, that's active regardless of the stats module itself?

File: projects/plugins/jetpack/modules/stats/class-jetpack-stats-dashboard-widget.php
Comment (jeherve) [inline projects/plugins/jetpack/modules/stats/class-jetpack-stats-dashboard-widget.php:83]: You'll probably want to keep the docblock for that filter.

File: projects/plugins/jetpack/modules/stats/class-jetpack-stats-dashboard-widget.php
Comment (jeherve) [inline projects/plugins/jetpack/modules/stats/class-jetpack-stats-dashboard-widget.php:79]: I wonder if it may make sense to be more verbose here, or in the docblock for the action. Why do we rely on just an action here? What do we expect to be hooked to that action? It may save folks some time in the future, when trying to understand why this is built that way (we don't just add a footer, we also rely on the action to hook the actual graph).

**Review by jeherve — COMMENTED** [review]
> we make a general "Jetpack Dashboard Widget" class that we can use to load whatever we want, ie akismet spam blocked, stats, etc, that's active regardless of the stats module itself? What do you think @jeherve ?

That sounds good to me, yeah. 👍

File: projects/plugins/jetpack/modules/stats/class-jetpack-stats-dashboard-widget.php
Comment (sdixon194) [reply projects/plugins/jetpack/modules/stats/class-jetpack-stats-dashboard-widget.php]: Doing some more testing, and I can't find a scenario where this is actually called. If we're not connected, the action isn't registered and the subsequent code block that checks for that action doesn't run. 

We could probably move it around, but I think we can just remove it. What do y'all think?

File: projects/plugins/jetpack/modules/stats/class-jetpack-stats-dashboard-widget.php
Comment (sdixon194) [reply projects/plugins/jetpack/modules/stats/class-jetpack-stats-dashboard-widget.php]: Took a stab at this.

**Comment by sdixon194** [comment]
I believe this is good for another review now that it's been moved to its own class at the Jetpack level. It was suspiciously easier than I thought, so I'm surely overlooking something.

File: projects/plugins/jetpack/class-jetpack-stats-dashboard-widget.php
Comment (samiff) [inline projects/plugins/jetpack/class-jetpack-stats-dashboard-widget.php:65]: It's not clear to me why we'd bother reordering the widget? Am I missing something, or can we just drop this entire block.

If users have reordered their widgets it sounds like it won't even do anything anyways:

> ...this only works for people who have never re-ordered their widgets. Once a user has done so their existing preferences will override this and they will have to move your widget to the top for it to stay there.
via https://developer.wordpress.org/apis/handbook/dashboard-widgets/#forcing-your-widget-to-the-top

Comment (sdixon194) [reply projects/plugins/jetpack/class-jetpack-stats-dashboard-widget.php:65]: Yeah I mentioned that [in this comment as well](https://github.com/Automattic/jetpack/pull/20676#discussion_r708630721). I can't find a scenario where the code is run anyway. I think we can nix it.

Comment (samiff) [reply projects/plugins/jetpack/class-jetpack-stats-dashboard-widget.php:65]: @sdixon194 yeah, go ahead and drop that block and I can give it a final look.

Comment (sdixon194) [reply projects/plugins/jetpack/class-jetpack-stats-dashboard-widget.php]: removed in 7a45952

**Review by samiff — DISMISSED** [review]

File: projects/plugins/jetpack/modules/stats/class-jetpack-stats-dashboard-widget.php
Comment (samiff) [reply projects/plugins/jetpack/modules/stats/class-jetpack-stats-dashboard-widget.php]: This block was removed, see: https://github.com/Automattic/jetpack/pull/20676#discussion_r711384296

**Review by samiff — APPROVED** [review]
Re-approving after merge mess 🙂

**Pull request merged**
